streamz/tests/test_kafka.py

```diff
@@ -291,11 +294,13 @@ def test_kafka_dask_checkpointing_sync_nodes(c, s, w1, w2):
     j1 = random.randint(0, 10000)
     ARGS1 = {'bootstrap.servers': 'localhost:9092',
             'group.id': 'streamz-test%i' % j1,
-            'enable.auto.commit': False}
+            'enable.auto.commit': False,
+            'auto.offset.reset': 'earliest'}
     j2 = j1 + 1
     ARGS2 = {'bootstrap.servers': 'localhost:9092',
             'group.id': 'streamz-test%i' % j2,
-            'enable.auto.commit': False}
+            'enable.auto.commit': False,
+            'auto.offset.reset': 'earliest'}
```

Comment:
We could do with some more tests, e.g., with different number of partitions and whether the user passes npartitions or not (and what if the number passed is larger or smaller than that set in kafka); and the offsets earliest/latest. 

This checkpointing test is I suppose a separate thing, but it would be nice to see all the combinations tested somewhere.

Reply:
> with different number of partitions

Yes, we need this test.

> whether the user passes npartitions or not

We already have this, correct? But I guess none of the tests have the `npartitions` parameter specified  (i.e., all of them use the new default value of `None`), which I think is ok. But yeah, we can add one with `npartitions` maybe specified to something >1. 

> and what if the number passed is larger or smaller than that set in kafka

Not sure I understand. The support for adding dynamic partitions on the fly would be a follow-up PR like we discussed (auto-refresh partitions every `N` cycles based on a user parameter).

 > offsets earliest/latest

These have been covered extensively, although they are spread across different tests. 

Should I add a test with `npartitions` specified and >1 for now? Or can it be in a separate PR with the adding partitions on the fly capability?

Reply:
I think certainly we should add npartitions=-1, 0, 1, 2, 4 as a set of tests where the number of real partitions in kafka is 2 (fixed). The first two should fail with ValueError, the middle one should get about half the data, and the last two should both get all the data, but the last one will have partitions that never receive anything. Does that sound right?

Reply:
For npartitions=-1/0, it doesn't enter the polling consumption [loop](https://github.com/chinmaychandak/streamz/blob/ccc7d428ab24939fe07b151a7469a335f853ef9b/streamz/sources.py#L512). 

For npartitions=1, yes, about half the data. 

For npartitions=2/4, yes, they get all the data, but in the last one, partitions 3 & 4 are skipped in every polling consumption loop because the `get_watermarks_offset` throws an [error](https://github.com/chinmaychandak/streamz/blob/ccc7d428ab24939fe07b151a7469a335f853ef9b/streamz/sources.py#L517). 

All of these are handled within `from_kafka_batched` gracefully IMO (we can add a check at the beginning itself that raises a ValueError if npartitions is <=0, if need be), so we might not need tests. However, I do think we need a test with `npartitions` specified explicitly with a value `>1`, and maybe it can be plugged inside the test with the auto-refresh capability.

What do you think?

Reply:
Added some tests, do have a look

Reply:
@martindurant, please do let me know if we need to add anything else to this to get it rolled in. If not, there's another PR for auto-refreshing partitions on the way! :)